### PR TITLE
[Admin] Admin UI showcase

### PR DESCRIPTION
**Pull request by elia**
## Summary

Now that we're getting into the nitty gritty of building the new UI with components we need.

This PR sets up component previews for SolidusAdmin and adds a basic [Lookbook](https://lookbook.build) setup to the sandbox. The reason lookbook wasn't added to SolidusAdmin directly is doesn't support being mounted at multiple paths with dedicated configurations. For that reason it's better to prioritize host application setting up their lookbook installation for storefront components if they wish to.

A basic component preview boilerplate is setup when generating components, mainly consisting in an `overview` and a `playground` preview scenarios.

The **playground** scenario will allow playing with dynamic params live from Lookbook, e.g. changing the color of a badge.

The **overview** scenario is intended to showcase the component in different modes (e.g. a button with all its states and options in a single matrix/table).
The `overview` can also act as a simple smoke test for the component and can act as basic spec coverage for the component that ensures is not breaking when provided with different options.

### Other approaches

Before settling to starting with the `overview` and `playground` scenarios I tried a bunch of different approaches using [Lookbook groups](https://lookbook.build/guide/previews/groups), including: 
- typing out all the scenarios manually: `def badge_with_green_color_and_m_size…`
- using a local macro `def self.add_scenario…` followed by `add_scenario color: :green, size: :m`
and other in-between approaches.

Eventually it was very time consuming and repetitive, even when using the `# !@macro` directive of YARD.
The solution I settled for is generating the overview template with a simple layout similar to the one used for Lookbook groups, and for the badge component using a simple table resembling the Figma designs I used to build it.

That seemed the best compromise that also minimized effort and duplication.

<details><summary>Details</summary>
<p>

```rb
  # @!macro [attach] with_color_and_size
  #   @!method $1_$2
  def self.with_color_and_size(color, size)
    define_method "#{color}_#{size}" do
      configurable(color: color, size: size)
    end
  end

  # @!group Graphite Light
  with_color_and_size :graphite_light, :s
  with_color_and_size :graphite_light, :m
  with_color_and_size :graphite_light, :l
  # @!endgroup

  # @!group Red
  with_color_and_size :red, :s
  with_color_and_size :red, :m
  with_color_and_size :red, :l
  # @!endgroup

  # @!group Green
  with_color_and_size :green, :s
  with_color_and_size :green, :m
  with_color_and_size :green, :l
  # @!endgroup

  # @!group Blue
  with_color_and_size :blue, :s
  with_color_and_size :blue, :m
  with_color_and_size :blue, :l
  # @!endgroup

  # @!group Black
  with_color_and_size :black, :s
  with_color_and_size :black, :m
  with_color_and_size :black, :l
  # @!endgroup

  # @!group Yellow
  with_color_and_size :yellow, :s
  with_color_and_size :yellow, :m
  with_color_and_size :yellow, :l
  # @!endgroup

  # # @!group Medium
  # with_color_and_size :graphite_light, :m
  # with_color_and_size :red, :m
  # with_color_and_size :green, :m
  # with_color_and_size :blue, :m
  # with_color_and_size :black, :m
  # with_color_and_size :yellow, :m
  # # @!endgroup

  # # @!group Small
  # with_color_and_size :graphite_light, :s
  # with_color_and_size :red, :s
  # with_color_and_size :green, :s
  # with_color_and_size :blue, :s
  # with_color_and_size :black, :s
  # with_color_and_size :yellow, :s
  # # @!endgroup
```

</p>
</details> 

## Screenshots

image
image

### ViewComponent previews

The original VC previews are still accessible at `http://localhost:3000/rails/view_components`

image

`http://localhost:3000/rails/view_components/solidus_admin/ui/badge/component/overview`

image

`http://localhost:3000/rails/view_components/solidus_admin/ui/badge/component/playground`

image

`http://localhost:3000/rails/view_components/solidus_admin/ui/badge/component/playground?name=Foo`

image

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [ ] I have written a thorough PR description.
- [ ] I have kept my commits small and atomic.
- [ ] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

### CODE REVIEW

File: bin/sandbox [hunk 1]
```diff
@@ -99,7 +99,11 @@ unbundled bin/rails generate solidus:install \
   --auto-accept \
   $@
 
+unbundled bundle add lookbook listen actioncable --group development,test
+ruby -e "File.write('config/routes.rb', File.read('config/routes.rb').gsub(/draw do/, %{draw do\n  mount Lookbook::Engine, at: '/lookbook' if Rails.env.development?}))"
```
Comment (rainerdema) [inline bin/sandbox:103]: Just a note on this, as we discussed during the live review, this could be moved from here probably 🤔
Even if to be honest, I don't have a better idea for the moment on how to do it differently.

**Review by rainerdema — APPROVED** [review]
This will be extremely useful for developing various UI components and for developers who want to do their own integration on the new Solidus admin! 😍
Looks great 👏

File: bin/sandbox [hunk 2]
```diff
@@ -99,7 +99,11 @@ unbundled bin/rails generate solidus:install \
   --auto-accept \
   $@
 
+unbundled bundle add lookbook listen actioncable --group development,test
```
Comment (waiting-for-dev) [inline bin/sandbox:102]: For some reason, it looks like the gems are not added to the sandbox's `Gemfile`.

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# @component "ui/badge"
+class SolidusAdmin::UI::Badge::ComponentPreview < ViewComponent::Preview
+  layout "solidus_admin/preview"
+  include SolidusAdmin::ContainerHelper
```
Comment (waiting-for-dev) [inline admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:6]: What do you think about moving it into a `SolidusAdmin::BaseComponentPreview` class? I guess we'll need those for all other previews...

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  # @param name [String]
+  def overview(name: "Label")
+    render_with_template(locals: { name: name, component: component("ui/badge") })
```
Comment (waiting-for-dev) [inline admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:10]: If we add the `BaseComponentPreview`, maybe we can extract that `component("ui/badge")` into a helper method `component`? We can do it later, though.

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+RSpec.describe SolidusAdmin::UI::Badge::Component, type: :component do
+  it "renders the overview preview" do
+    render_preview(:overview)
```
Comment (waiting-for-dev) [inline admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: I tried it locally and it's failing with:

```ActionView::MissingTemplate:
       Missing template layouts/solidus_admin/preview with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby]}.
     
       Searched in:
         * "/home/solidus_user/app/admin/spec/components/previews"
         * "/home/solidus_user/app/admin/spec/dummy/app/views"
         * "/home/solidus_user/gems/gems/view_component-3.2.0/app/views"
         * "/home/solidus_user/gems/gems/railties-7.0.4.3/lib/rails/templates"
     
       Did you mean?  solidus_admin/ui/badge/component_preview/overview
                      view_components/preview
                      view_components/previews
                      layouts/application
                      rails/info/properties
                      solidus_admin/ui/badge/component_preview
```

For some reason, it looks like CI is only running `core` & `backend` for this PR :thinking: It probably has nothing to do with the work here, so we need to investigate...

**Review by waiting-for-dev — COMMENTED** [review]
Thanks, @elia! I love having this feature, and I think the overview/playground structure can work well. Left some comments, and I think there are a few issues that need to be fixed.

File: admin/app/components/solidus_admin/ui/badge/component.rb [hunk 6]
```diff
@@ -12,12 +12,18 @@ class SolidusAdmin::UI::Badge::Component < SolidusAdmin::BaseComponent
     yellow: 'text-yellow bg-[#feecd4]',
   }.freeze
 
-  def initialize(name:, color: :graphite_light)
+  SIZES = {
+    s: 'leading-[16px] px-[8px] py-[2px] text-[12px] font-[500]',
+    m: 'leading-[20px] px-[12px] py-[2px] text-[14px] font-[500]',
+    l: 'leading-[24px] px-[12px] py-[2px] text-[16px] font-[500]',
```
[the remark this answers is not in the document]
Comment (elia) [reply admin/app/components/solidus_admin/ui/badge/component.rb:18]: Yes, I would consider that out of the scope of this PR. I added this component and product index to a list of think to be checked in #5186 👍

File: admin/lib/solidus_admin/configuration.rb [hunk 7]
```diff
@@ -27,6 +27,7 @@ class Configuration < Spree::Preferences::Configuration
       SolidusAdmin::Engine.root.join("app/assets/javascripts/**/*.js"),
       SolidusAdmin::Engine.root.join("app/views/**/*.erb"),
       SolidusAdmin::Engine.root.join("app/components/**/*.{rb,erb,js}"),
+      SolidusAdmin::Engine.root.join("spec/components/previews/**/*.{rb,erb,js,html,md}"),
```
[the remark this answers is not in the document]
Comment (elia) [reply admin/lib/solidus_admin/configuration.rb:30]: This was to easily style surrounding html in the overview scenario or for potential other scenarios in which we want to show the component in a specific context.

Of course some of those styles might not be strictly needed in the actual admin, but I think it's just a few classes and it's not worth having a separate TW build or using inline style attributes.

File: bin/sandbox (hunk 2 above)
Comment (elia) [reply bin/sandbox:102]: Is it a silent error or you get anything more specific? 
Also I suggest checking if it's the latest version of bundler, just to be sure.

File: bin/sandbox (hunk 1 above)
Comment (elia) [reply bin/sandbox:103]: I'll try to move both operations to the solidus_admin install generator 👍

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb (hunk 3 above)
Comment (elia) [reply admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:6]: This bothers me as well, 😞 and I tried, but for some reason the preview broke, after a while I stopped spending time on it and just added the two lines to the generator. I think over time we might give it another crack and see if we succeed. 

I can try again if you think this is needed right away.

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb (hunk 4 above)
Comment (elia) [reply admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:10]: Good call! You think would be worth it having a local helper with some duplication assuming we won't be able to extract the superclass?

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (elia) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: I tried locally and it works fine, maybe we used different commands?

```
solidus_admin:elia/admin/ui-showcase ⤑ bin/rspec admin/spec/components/solidus_admin/ui/badge/component_spec.rb                                   ~/C/N/solidus_admin
$ cd admin; bundle exec rspec /Users/elia/Code/Nebulab/solidus_admin/admin/spec/components/solidus_admin/ui/badge/component_spec.rb; cd -

Randomized with seed 64336
.

Finished in 0.24157 seconds (files took 2.02 seconds to load)
1 example, 0 failures

Randomized with seed 64336
```

File: admin/lib/solidus_admin/configuration.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (waiting-for-dev) [reply admin/lib/solidus_admin/configuration.rb:30]: I see. It's probably an acceptable compromise. However, we probably only need the `erb` extension now.

File: bin/sandbox (hunk 2 above)
Comment (waiting-for-dev) [reply bin/sandbox:102]: Very silent :slightly_smiling_face: Do you get it added in your sandbox?

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb (hunk 3 above)
Comment (waiting-for-dev) [reply admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:6]: It's not needed now, but it probably will be needed for the next one :slightly_smiling_face: We can work on it separately, though.

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb (hunk 4 above)
Comment (waiting-for-dev) [reply admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:10]: I'm not sure I follow... if we had the `BaseComponentPreview` (which it looks it's not that easy) we should be able to extract the component from the preview class name, right? `class SolidusAdmin::UI::Badge::ComponentPreview` => `ui/badge`

Comment (elia) [reply admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:10]: Yes, if we had the BaseComponentPreview that would be super easy, I was wondering if would make sense to have it in every preview class while we don't have the base class. Sorry about the confusion.

File: bin/sandbox (hunk 2 above)
Comment (elia) [reply bin/sandbox]: It is now part of the install generator, does it work for you now?

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (elia) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: > For some reason, it looks like CI is only running `core` & `backend` for this PR 🤔 It probably has nothing to do with the work here, so we need to investigate...

@waiting-for-dev I was fooled by CircleCI as well at first, you can find `admin` in one of the other parallel runs, `bin/build-ci` tries to split the work among parallel runs.

image

Comment (waiting-for-dev) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: That's really weird. @rainerdema, @kennyadsl, could you try it and confirm who's seeing ghosts here? :slightly_smiling_face: I tried once more, and it fails for me...`q

File: admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb (hunk 4 above)
Comment (waiting-for-dev) [reply admin/spec/components/previews/solidus_admin/ui/badge/component_preview.rb:10]: Yeah, that works if the base class is not an option.

File: bin/sandbox (hunk 2 above)
Comment (kennyadsl) [reply bin/sandbox]: I tried it yesterday and it was there.

File: admin/lib/generators/solidus_admin/install/install_generator.rb [hunk 8]
```diff
@@ -24,6 +24,19 @@ def ignore_tailwind_build_files
...
+        # The dummy app has no Gemfile
+        return unless in_root { File.exist?("Gemfile") }
+
+        gem_group :development, :test do
+          gem "lookbook"
+          gem "listen"
+          gem "actioncable"
+        end
+
+        route "mount Lookbook::Engine, at: '/lookbook' if Rails.env.development?"
```
Comment (waiting-for-dev) [inline admin/lib/generators/solidus_admin/install/install_generator.rb:38]: That will affect every `solidus_admin` installation unless I'm missing something. What about controlling through an environment variable and we set it in the sandbox script?

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (kennyadsl) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: Let me try again!

File: bin/sandbox (hunk 2 above)
Comment (waiting-for-dev) [reply bin/sandbox]: I tried now and it works, but I left a comment about the approach.

File: admin/lib/generators/solidus_admin/install/install_generator.rb (hunk 8 above)
Comment (kennyadsl) [reply admin/lib/generators/solidus_admin/install/install_generator.rb:38]: That's a good idea!

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (elia) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: @rainerdema had the same issue, looks like there's a fix for that issue in ViewComponent 3.3, I'll bump the dependency in solidus_admin.gemspec 👍 

Probably me and @kennyadsl already have 3.3.

File: admin/lib/generators/solidus_admin/install/install_generator.rb (hunk 8 above)
Comment (elia) [reply admin/lib/generators/solidus_admin/install/install_generator.rb:38]: I think it's fine to install it anyway, and the starter should use it as well, that said I added a class option with the default dependant on the `SOLIDUS_ADMIN_LOOKBOOK` env variable, and changed the sandbox script accordingly.

File: admin/lib/generators/solidus_admin/install/install_generator.rb [hunk 9]
```diff
@@ -24,6 +26,21 @@ def ignore_tailwind_build_files
       def build_tailwind
         rake "solidus_admin:tailwindcss:build"
       end
+
+      def install_lookbook
+        return unless options[:lookbook]
+
+        # The dummy app has no Gemfile
+        return unless in_root { File.exist?("Gemfile") }
```
Comment (waiting-for-dev) [inline admin/lib/generators/solidus_admin/install/install_generator.rb:34]: Nit, but maybe we can now remove these lines, as we won't provide `SOLIDUS_ADMIN_LOOKBOOK` variable for the dummy app.

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (waiting-for-dev) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: I bumped the dependency to 3.3, but I'm still experiencing the problem. Is it the same for you, @rainerdema? :thinking:

File: admin/lib/generators/solidus_admin/install/install_generator.rb [hunk 10]
```diff
@@ -3,7 +3,9 @@
 module SolidusAdmin
   module Generators
     class InstallGenerator < Rails::Generators::Base
-      source_root File.expand_path('templates', __dir__)
+      class_option :lookbook, type: :boolean, default: !!ENV['SOLIDUS_ADMIN_LOOKBOOK'], banner: 'Install Lookbook for component previews'
```
Comment (kennyadsl) [inline admin/lib/generators/solidus_admin/install/install_generator.rb:6]: I tried to run ` bin/rails g solidus_admin:install --help`

```
Running Rails from: /Users/kennyadsl/Code/nebulab/solidus/sandbox
Usage:
  rails generate solidus_admin:install [options]

Options:
  [--skip-namespace], [--no-skip-namespace]                              # Skip namespace (affects only isolated engines)
  [--skip-collision-check], [--no-skip-collision-check]                  # Skip collision check
  [--lookbook=Install Lookbook for component previews], [--no-lookbook]  # Indicates when to generate lookbook

Runtime options:
  -f, [--force]                    # Overwrite files that already exist
  -p, [--pretend], [--no-pretend]  # Run but do not make any changes
  -q, [--quiet], [--no-quiet]      # Suppress status output
  -s, [--skip], [--no-skip]        # Skip files that already exist

Description:
    Create solidus admin files for install generator.
```

And I think there's something off with the `lookbook` option. It's setting the banner in the value of the option, which doesn't seem correct to me.

**Review by kennyadsl — COMMENTED** [review]
I think there's something to review in the install generator options, and there's a commit to rebase. Other than that, I'm good to go, looking forward to approval and merging!

File: admin/lib/generators/solidus_admin/install/install_generator.rb (hunk 10 above)
Comment (elia) [reply admin/lib/generators/solidus_admin/install/install_generator.rb:6]: @kennyadsl good catch, I copied it from the core install generator, but maybe being a different superclass the syntax is different, fixing it 👍

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (elia) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: @waiting-for-dev the dependency must be updated on the root Gemfile, not the sandbox one, I thought it was worth mentioning since this tricked me at first.

Comment (rainerdema) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: @waiting-for-dev I can confirm that I was experiencing the same exception yesterday.
But after I switched from version 3.2.0 to 3.3.0, the error ceased to occur 🤔

The fix for the problem released in 3.3.0 seems to be this specifically:
https://github.com/ViewComponent/view_component/pull/1777

File: admin/lib/generators/solidus_admin/install/install_generator.rb [hunk 11]
```diff
@@ -24,6 +26,18 @@ def ignore_tailwind_build_files
...
+      def install_lookbook
+        return unless options[:lookbook]
+
+        gem_group :development, :test do
+          gem "lookbook"
+          gem "listen"
+          gem "actioncable"
+        end
+
+        route "mount Lookbook::Engine, at: '/lookbook' if Rails.env.development?"
```
Comment (kennyadsl) [inline admin/lib/generators/solidus_admin/install/install_generator.rb:39]: @elia can you help me understand why we need `if Rails.env.development?"` if we also have the ENV variable?

**Review by kennyadsl — COMMENTED** [review]
👍 left a last, non-blocking thing. Thanks!

**Review by kennyadsl — APPROVED** [review]
Approving because the last thing was not blocking 😅

File: admin/lib/generators/solidus_admin/install/install_generator.rb (hunk 11 above)
Comment (elia) [reply admin/lib/generators/solidus_admin/install/install_generator.rb:39]: That is how lookbook suggests to add it to routes, the same goes for the bundler groups the gems will belong to. Also the env variable is about whether we'll install it, not for when it should be active.

I think it's good to stick to lookbook's instructions when we install it to the host app.

File: admin/spec/components/solidus_admin/ui/badge/component_spec.rb (hunk 5 above)
Comment (waiting-for-dev) [reply admin/spec/components/solidus_admin/ui/badge/component_spec.rb:7]: Bumping to v3.3.0 alone didn't fix my issue, but after re-generating my `Gemfile.lock` file the issue is gone.

**Review by waiting-for-dev — APPROVED** [review]
:rocket:
